src/machine/machine-varlink.c

```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
+        }
+}
+
+inline static const char** machine_open_polkit_details(MachineOpenMode mode, const char *machine_name, const char *user, const char *path, const char *command_line) {
+        assert(machine_name);
+
+        switch (mode) {
+                case MACHINE_OPEN_MODE_TTY:
+                        return (const char**) STRV_MAKE("machine", machine_name);
+                case MACHINE_OPEN_MODE_LOGIN:
+                        return (const char**) STRV_MAKE("machine", machine_name, "verb", "login");
+                case MACHINE_OPEN_MODE_SHELL:
+                        assert(user);
+                        assert(path);
+                        assert(command_line);
+
+                        return (const char**) STRV_MAKE(
+                                        "machine", machine_name,
+                                        "verb", "shell",
+                                        "user", user,
+                                        "program", path,
+                                        "command_line", command_line);
+                default:
+                        assert_not_reached();
+        }
+}
+
+int vl_method_open(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dispatch_table[] = {
+                VARLINK_DISPATCH_MACHINE_LOOKUP_FIELDS(MachineOpenParameters),
+                { "mode", SD_JSON_VARIANT_STRING, json_dispatch_machine_open_mode, offsetof(MachineOpenParameters, mode), 0 },
+                { "user", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, user), 0 },
+                { "path", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, path), 0 },
+                { "args", SD_JSON_VARIANT_ARRAY,  sd_json_dispatch_strv,           offsetof(MachineOpenParameters, args), 0 },
+                { "env",  SD_JSON_VARIANT_ARRAY,  sd_json_dispatch_strv,           offsetof(MachineOpenParameters, env),  0 },
+                VARLINK_DISPATCH_POLKIT_FIELD,
+                {}
+        };
+
+        Manager *manager = ASSERT_PTR(userdata);
+        _cleanup_close_ int master = -EBADF;
+        _cleanup_(machine_open_paramaters_done) MachineOpenParameters p = { .pidref = PIDREF_NULL };
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_free_ char *pty_name = NULL, *command_line = NULL;
+        _cleanup_strv_free_ char **args = NULL;
+        const char *user = NULL, *path = NULL; /* gcc complains about uninitialized variables */
+        Machine *machine;
+        int r;
+
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, dispatch_table, &p);
+        if (r != 0)
+                return r;
+
+        if (p.mode == MACHINE_OPEN_MODE_SHELL) {
+                if (!isempty(p.user) && !valid_user_group_name(p.user, VALID_USER_RELAX))
```

Comment:
i don't think we should use isempty() here. isempty() is intended for cases when the empty string and unset shall be treated the same. We use that often in D-BUS IPC code since D-Bus has no "nullable" concept, and you cannot emit fields, hence we often use the empty string as "niche" value for "unset". But in a json/varlink world that's unnecessary, we have a nullable concept. hence i think we should refuse empty usernames, but allow unset usernames.

tldr. change `!isempty(p.user)` to just `p.user`